atomic_reactor/utils/imageutil.py

```diff
@@ -5,18 +5,95 @@
 This software may be modified and distributed under the terms
 of the BSD license. See the LICENSE file for details.
 """
-# OSBS2 TBD
 
+import functools
+from typing import Optional, Union
 
-def get_inspect_for_image(image):
-    # util.get_inspect_for_image(image, registry, insecure=False, dockercfg_path=None)
-    # or use skopeo
-    # insecure = self.pull_registries[base_image.registry]['insecure']
-    # dockercfg_path = self.pull_registries[base_image.registry]['dockercfg_path']
-    # self._base_image_inspect =\
-    #     atomic_reactor.util.get_inspect_for_image(base_image, base_image.registry, insecure,
-    # dockercfg_path)
-    return {}
+from osbs.utils import ImageName
+
+from atomic_reactor import config
+from atomic_reactor import util
+from atomic_reactor.types import ImageInspectionData
+
+
+def inspectable(image: Union[str, ImageName]) -> bool:
+    """Check if we should expect the image to be inspectable."""
+    im = str(image)
+    return not (util.base_image_is_scratch(im) or util.base_image_is_custom(im))
+
+
+class ImageUtil:
+    """Convenience class for working with images relevant to the build process.
+
+    Supports e.g. inspecting the base image and other parent images.
+    """
+
+    def __init__(self, dockerfile_images: util.DockerfileImages, conf: config.Configuration):
+        """Init an ImageUtil.
+
+        :param dockerfile_images: information about the image references in the Dockerfile
+        :param conf: atomic-reactor configuration
+        """
+        self._dockerfile_images = dockerfile_images
+        self._conf = conf
+
+    def set_dockerfile_images(self, dockerfile_images: util.DockerfileImages) -> None:
+        """Set a new dockerfile_images instance."""
+        self._dockerfile_images = dockerfile_images
+
+    def get_inspect_for_image(
+        self, image: Union[str, ImageName], platform: Optional[str] = None
+    ) -> ImageInspectionData:
+        """Inspect an image. Should mainly be used to inspect parent images.
+
+        The image must be inspectable, passing a non-inspectable image is an error.
+
+        The result is cached, this method will not query the registry more than once
+        (if successful) for the same image reference (within one Python process).
+
+        :param image: The image to inspect
+        :param platform: Optionally, inspect the base image for a specific platform.
+            This can be either the platform name (e.g. x86_64) or the GOARCH name (amd64).
+        """
+        if not inspectable(image):
+            raise ValueError(f"{image!r} is not inspectable")
```

Comment:
mmm I wonder if we shouldn't just return {}

I mean code shouldn't try to inspect scratch/custom base images anyway, but still

Reply:
Hmm, yeah that would be more consistent with how `base_image_inspect` behaves. On the other hand, unlike the `base_image_inspect` method, when calling this one the caller knows what image they are passing as an argument so maybe it should be up to them to make sure the image is valid.

I'm fine with either, I'd like to know what others think